File: JavaScript/1804-Implement Trie II (Prefix Tree).js
var TrieNode = function(){
    this.children = new Map();
    this.prefix = 0;
    this.count = 0;
}


var Trie = function() {
    this.root = new TrieNode();
};

/** 
 * @param {string} word
 * @return {void}
 */
Trie.prototype.insert = function(word) {
    let curr = this.root;
    for (let w of word){
        if(!curr.children.has(w)){
            curr.children.set(w, new TrieNode())
        }
        curr = curr.children.get(w);
        curr.prefix+=1;
    }
    curr.count += 1;
};

/** 
 * @param {string} word
 * @return {number}
 */
Trie.prototype.countWordsEqualTo = function(word) {
    let curr = this.root;
    for (let w of word){
        if(!curr.children.has(w)){
            return 0
        }
        curr = curr.children.get(w);
    }
    return curr.count;
};

/** 
 * @param {string} prefix
 * @return {number}
 */
Trie.prototype.countWordsStartingWith = function(prefix) {
    let curr = this.root;
    for (let w of prefix){
        if(!curr.children.has(w)){
            return 0
        }
        curr = curr.children.get(w);
    }
    return curr.prefix;    
};

/** 
 * @param {string} word
 * @return {void}
 */
Trie.prototype.erase = function(word) {
    let curr = this.root;
    for (let w of word){
        if(!curr.children.has(w)){
            return
        }
        let nxt = curr.children.get(w)
        nxt.prefix -= 1
        if (nxt.prefix === 0){
            curr.children.delete(w);
            return
        }
        curr = nxt;
    }
    curr.count -= 1;
};

/** 
 * Your Trie object will be instantiated and called as such:
 * var obj = new Trie()
 * obj.insert(word)
 * var param_2 = obj.countWordsEqualTo(word)
 * var param_3 = obj.countWordsStartingWith(prefix)
 * obj.erase(word)
 */

File: C++/1804-Implement Trie II (Prefix Tree).cpp
class Trie {
public:
    struct TrieNode {
        TrieNode* ch[26];
        int cnt;
        int isEnd;
        explicit TrieNode():cnt(0), isEnd(0){
            std::fill(ch, ch +26 , nullptr);
        }
        virtual ~TrieNode(){ // this makes code even faster
            for(TrieNode* e : ch) delete e;
        }
    };
    TrieNode* root;
    Trie() {
        root = new TrieNode();
    }
    
    void insert(string word) {
        TrieNode* node = root;
        for (char w:word) {
            if (node->ch[w-'a'] == nullptr) {
                node->ch[w-'a'] = new TrieNode();
            }
            node = node->ch[w-'a'];
            node->cnt++;
        }
        node->isEnd++;
    }
    
    int countWordsEqualTo(string word) {
        TrieNode* node = root;
        for (char w:word) {
            if (node->ch[w-'a'] == nullptr) {
                node->ch[w-'a'] = new TrieNode();
            }
            node = node->ch[w-'a'];
        }
        return node->isEnd;       
    }
    
    int countWordsStartingWith(string prefix) {
        TrieNode* node = root;
        for (char w:prefix) {
            if (node->ch[w-'a'] == nullptr) {
                return 0;
            }
            node = node->ch[w-'a'];
        }
        return node->cnt;        
    }
    
    void erase(string word) {
        TrieNode* node = root;
        for (char w:word) {
            node = node->ch[w-'a'];
            --node->cnt; 
        }
        --node->isEnd;
    }
};
